Task17 jaxb

andresmadeddie opened the pull request:
Create JAXB annotations on all classes inside the model package
Parse XML using JAXB

andresmadeddie committed:
Add JAXB dependency to pom files
Added @XmlRootElement to all classes in jdbc directory
Added annotations to the model class

andresmadeddie committed:
Remove annotations in all files of JDBC directory

andresmadeddie committed:
Modify all files on model from Property to Fields and each element on it accordingly

andresmadeddie committed:
Add DateAdapter class
Implement DateAdapter on promotion and orders model annotations
change to XmlAttribute id fields on all archives who have it inside the model folder

andresmadeddie committed:
update

andresmadeddie committed:
Add JAXB impl repository

andresmadeddie committed:
Add JAXB impl repository
Create folder with files to test JABX functionality
Create MarshOrUnmash Class and use its methods to marsh and unmarsh all classes in the Model directory at the main class

andresmadeddie committed:
Add JAXB impl repository
Create folder with files to test JABX functionality
Create MarshOrUnmash Class and use its methods to marsh and unmarsh all classes in the Model directory at the main class

andresmadeddie committed:
delete unused Marsh files and move MarshOrUnMarsh class to the JAXB folder

andresmadeddie committed:
minor fixes

andresmadeddie committed:
minor fixes

jinocencio-solvd commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,106 @@
+
+    // Get all Classes
+    public static List<Class<?>> getClasses(String packageName) {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
```
I considered this approach too! nice implementation

jinocencio-solvd commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,106 @@
+                    for (File f : files) {
+                        String fileName = f.getName();
+                        if (fileName.endsWith(".class")) {
+                            String className = packageName + "." + fileName.substring(0, fileName.length() - 6);
```
6 is a magic number here. Can you describe what 6 represents ?

andresmadeddie replied:
> 6 is a magic number here. Can you describe what 6 represents ?

.class

jinocencio-solvd replied:
I'd recommend using this in your code and avoid magic numbers so that it is more readable and easy to understand. For example you can declare a String called classExt = ".class", use it in your if condition, and call classExt.length() in place of 6

jinocencio-solvd commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,106 @@
+            // Marshal the instance to XML and save it to the file
+            marshaller.marshal(instance, new File(fileName));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
why catch exception to throw RuntimeException?

jinocencio-solvd commented on src/main/java/com/solvd/test/BigBang.java:
```diff
@@ -0,0 +1,35 @@
+package com.solvd.test;
```
does this package still belong in your repo ?

ankochergin commented on src/main/java/com/solvd/test/Test.java:
```diff
@@ -0,0 +1,50 @@
+            System.out.println(bigBang);
+
+
+        } catch (JAXBException e) {
```
```suggestion
            System.out.println(bigBang);
        } catch (JAXBException e) {
```

ankochergin commented on src/main/java/com/solvd/test/BigBang.java:
```diff
@@ -0,0 +1,35 @@
+@XmlRootElement
+public class BigBang {
+    private String name;
+    private String last_name;
```
```suggestion
    private String lastName;
```

ankochergin commented on src/main/java/com/solvd/sql/model/Supplier.java:
```diff
@@ -1,9 +1,21 @@
+@XmlRootElement(name = "supplier")
 public class Supplier {
+
+    @XmlAttribute(name = "Id")
```
please use same naming conventions for all xml nodes (node name starts with lowercase character)

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.Enumeration;
+import java.util.List;
+
+public class MarshOrUnMarsh {
```
```suggestion
public class JAXBUtils {
```

ankochergin commented on src/main/java/com/solvd/sql/model/Orders.java:
```diff
@@ -1,11 +1,26 @@
+
+    @XmlElement(name = "order_date")
+    @XmlJavaTypeAdapter(DateAdapter.class)
     private Date order_date;
```
```suggestion
    private Date orderDate;
```

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,106 @@
+            unmarshall(clazz);
+        }
+    }
+
```
I like that you decided to implement solution to marshal and unmarshal different types using 2 methods! currently it creates new object using reflection inside of the method and marshals an empty object. What of I'd like to marshal/unmarshal some object from main? Could you please make not `Class<?> clazz` as input parameter, but an object/instance of any class from your hierarchy? We should be able to write any object from your main to an xml file
The same for reading objects, method unmarshal should return an object which we will be able to use in our code

andresmadeddie replied:
I modified the code so march and unMarch methods could get parameters ('object' or 'file path' respectively) from the main class.

andresmadeddie committed:
eliminate test folder and its content
change name unmarshItAll to unMarshItAll
Optimize redundant input for connection method in the connection pool class

andresmadeddie committed:
Change marshall method to get one object parameter

andresmadeddie committed:
change name of method from unmarshall to UnMarshall

andresmadeddie committed:
modify marshall and unmarshall methods to accept a path-file or an object. Modify  marshall and unmarshal all methods and the main class according
Create new method and class to extract tags from xml

andresmadeddie committed:
Remove Simple class and other minor fixes

ankochergin commented on src/main/java/com/solvd/util/XMLTagExtractor.java:
```diff
@@ -0,0 +1,33 @@
+public class XMLTagExtractor {
+    public static String fileExtractor(String filePath) {
+
+        String extractedString = "";
```
```suggestion
public class XMLTagExtractor {

    public static String fileExtractor(String filePath) {
        String extractedString = "";
```

ankochergin commented on src/main/java/com/solvd/util/XMLTagExtractor.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.regex.Pattern;
+
+public class XMLTagExtractor {
+    public static String fileExtractor(String filePath) {
```
if I got correctly it extracts main node to get the class name. if you method returns something better to start method name with `get`
lets rename it to `getMainNodeName()` or suggest other ideas

ankochergin commented on src/main/java/com/solvd/sql/model/Supplier.java:
```diff
@@ -1,9 +1,21 @@
+    @XmlElement(name = "TaxNumber")
     private String taxNumber;
+
+    @XmlElement(name = "Phone")
```
```suggestion
    @XmlAttribute(name = "id")
    private int id;

    @XmlElement(name = "supplierName")
    private String supplierName;

    @XmlElement(name = "taxNumber")
    private String taxNumber;

    @XmlElement(name = "phone")
```

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,105 @@
+                throw new RuntimeException(e);
+            }
+        }
+    }
```
I think we don't need this method anymore. I don't think that we have a reason to write an xml file without properly filled data

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,105 @@
+            String fileName = "src/main/resources/XMLFiles/" + clazz.getSimpleName() + ".xml";
+            System.out.println(unMarshall(fileName));
+        }
+    }
```
```suggestion
    public static void printAllXMLFiles() {
        String packageName = "com.solvd.sql.model";
        List<Class<?>> classes = getClasses(packageName);

        for (Class<?> clazz : classes) {
            String fileName = "src/main/resources/XMLFiles/" + clazz.getSimpleName() + ".xml";
            System.out.println(unMarshall(fileName));
        }
    }
```

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,105 @@
+        List<Class<?>> classes = getClasses(packageName);
+
+        for (Class<?> clazz : classes) {
+            String fileName = "src/main/resources/XMLFiles/" + clazz.getSimpleName() + ".xml";
```
path to the files, package and `.xml` strings are used in several methods, please extract them to final variables

ankochergin replied:
also rename `XMLFiles` to `xmloutput`

ankochergin commented on src/main/java/com/solvd/sql/jaxb/MarshOrUnMarsh.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.Enumeration;
+import java.util.List;
+
+public class MarshOrUnMarsh {
```
```suggestion
public class JAXBUtils {
```

andresmadeddie committed:
Change name form Id to id
Change name of class from MarshOrUnMarsh  to  JAXBUtils
Rename order_date to orderDate
Add line between class and first method in XMLExtractor
Rename fileExtractor to getMainNodeName()
Correct names of fields in Supplier
Remove marshItAll method
Rename unMarshItAll to printAllXMLFiles
Add enum with path variables
Rename XMLFiles to XMLFolder

ankochergin commented on src/main/java/com/solvd/util/XMLExtractor.java:
```diff
@@ -0,0 +1,34 @@
+
+    public static String getMainNodeName(String filePath) {
+
+        String extractedString = "";
```
```suggestion
    public static String getMainNodeName(String filePath) {
        String extractedString = "";
```

andresmadeddie committed:
minor fix
